concourse/client/util.py

```diff
@@ -122,3 +152,159 @@ def jobs_not_triggered(
+                        )).isoformat(),
+                        comment="pinned by technical user, please do not unpin manually"
+                    )
+                    concourse_api.pin_comment(
+                        pipeline_name=resource.pipeline_name(),
+                        resource_name=resource.name,
+                        comment=json.dumps(dataclasses.asdict(pin_comment)),
+                    )
+                    warning(
+                        f'job {job.name} for resource version {resource_version.version()} '
+                        f'could not be triggered. Will retry {retries} more time(s)'
+                    )
+                return
+        except PinningFailedError as e:
+            warning(e)
+            retries -= 1
+        except PinningUnnecessaryError as e:
+            # already beeing processed by another thread, nothing to do
+            info(e)
+            return
+
+    warning(
+        f'job {job.name} for resource version {resource_version.version()} '
+        f'could not be triggered. Giving up after {max_retries} retries'
+    )
+
+
+@contextmanager
+def pin_resource(
+    resource: PipelineResource,
+    resource_version: ResourceVersion,
+    concourse_api,
+):
+    _ensure_resource_unpinned(
+        resource=resource,
+        resource_version=resource_version,
+        concourse_api=concourse_api,
+        retries=10,
+        sleep_time_seconds=5,
+    )
+
+    try:
+        pin_comment = PinComment(
+            pin_timestamp=datetime.now().isoformat(),
+            version_data=resource_version.version(),
+            next_retry=(datetime.now() + timedelta(minutes=5)).isoformat(),
+            comment="pinned by technical user, please do not unpin manually"
+        )
+        concourse_api.pin_and_comment_resource_version(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+            resource_version_id=resource_version,
+            comment=json.dumps(dataclasses.asdict(pin_comment)),
+        )
+        yield
+    finally:
+        concourse_api.unpin_resource(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+        )
+
+
+def _ensure_resource_unpinned(
+    resource: PipelineResource,
+    resource_version: ResourceVersion,
+    concourse_api,
+    retries: int=10,
+    sleep_time_seconds: int=5,
+):
+    while retries >= 0:
+        cc_resource = concourse_api.resource(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+        )
+
+        if not cc_resource.is_pinned():
+            return
+
+        pin_comment_data = cc_resource.pin_comment()
+        try:
+            pin_comment = PinComment(**json.loads(pin_comment_data))
+            if pin_comment.version_data == resource_version.version():
+                raise PinningUnnecessaryError("Resource is already pinned for the same version")
```

Comment:
not sure I agree with this to be signalled as an error;

also: we should at least check whether the lock has become stale (i.e. timestamp too old)